File: reference/python/stiger_decode.py
"""Reference decoder for the Stiger wire format.

This file is the *canonical* example implementation of the spec in
`spec/open-v1.md` and `spec/stealth-v3.md`. It deliberately uses standard
library + widely-audited primitives (PyCA `cryptography`, Pillow), no Stiger
code, no clever tricks. If you can read this file you can read every Stiger
sticker you have the password for, and prove to yourself the binary on your
phone is not doing anything else.

Usage:

    python stiger_decode.py path/to/sticker.png
    python stiger_decode.py path/to/sticker.png --password 'hunter2'

Exit status:

    0  decoded successfully
    1  no payload / wrong password / corrupt
    2  bad arguments
"""

from __future__ import annotations

import argparse
import hashlib
import hmac
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from PIL import Image
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


# ---------------------------------------------------------------------------
# Constants — must match spec/open-v1.md and spec/stealth-v3.md.
# ---------------------------------------------------------------------------

OPEN_MAGIC = b"STEG"
OPEN_VERSION = 0x01
OPEN_HEADER_SIZE = 9

STEALTH_VERSION = 0x03
STEALTH_HEADER_SIZE = 22
STEALTH_LEN_PREFIX = 2

GCM_NONCE_LEN = 12
GCM_TAG_LEN = 16
GCM_OVERHEAD = GCM_NONCE_LEN + GCM_TAG_LEN

MAX_PAYLOAD_BYTES = 2048

PBKDF2_ITER = 600_000
SALT_LEN = 16

PERM_KDF_SALT = b"stiger-perm-v3-salt"
PERM_INFO = b"stiger-permutation-v3"
HEADER_INFO = b"stiger-v3-header"


# ---------------------------------------------------------------------------
# LSB transport layer (spec/lsb-layer.md).
# ---------------------------------------------------------------------------


@dataclass
class Slots:
    """Eligible (alpha == 255) pixel byte-offsets, row-major."""

    blue_offsets: list[int]      # offset of the blue byte in raw RGBA buffer
    raw: bytes                   # full RGBA buffer

    @property
    def n(self) -> int:
        return len(self.blue_offsets)

    @property
    def available_bytes(self) -> int:
        return self.n // 8


def load_slots(png_path: Path) -> Slots:
    img = Image.open(png_path).convert("RGBA")
    raw = img.tobytes()
    blue_offsets: list[int] = []
    for i in range(0, len(raw), 4):
        if raw[i + 3] == 0xFF:
            blue_offsets.append(i + 2)
    return Slots(blue_offsets=blue_offsets, raw=raw)


def read_bytes(slots: Slots, count: int, order: Optional[list[int]] = None,
               start_slot: int = 0) -> bytes:
    """Read `count` bytes (MSB-first within each byte) from the LSB stream.

    `order[i]` is the eligible-pixel index that holds logical bit `i`.
    `order=None` means row-major (logical i -> eligible pixel i).
    """
    out = bytearray(count)
    for byte_index in range(count):
        b = 0
        for bit in range(8):
            slot = start_slot + byte_index * 8 + bit
            if slot >= slots.n:
                break
            logical = order[slot] if order is not None else slot
            blue = slots.raw[slots.blue_offsets[logical]]
            b = (b << 1) | (blue & 1)
        out[byte_index] = b
    return bytes(out)


# ---------------------------------------------------------------------------
# Open mode (spec/open-v1.md).
# ---------------------------------------------------------------------------


def decode_open(slots: Slots) -> bytes:
    if slots.available_bytes < OPEN_HEADER_SIZE:
        raise ValueError("no payload found")

    header = read_bytes(slots, OPEN_HEADER_SIZE)
    if header[0:4] != OPEN_MAGIC:
        raise ValueError("no payload found")

    length = int.from_bytes(header[5:9], "big")
    if length == 0 or length > MAX_PAYLOAD_BYTES:
        raise ValueError("no payload found")
    if OPEN_HEADER_SIZE + length > slots.available_bytes:
        raise ValueError("payload truncated")

    return read_bytes(slots, length, start_slot=OPEN_HEADER_SIZE * 8)


# ---------------------------------------------------------------------------
# Stealth mode (spec/stealth-v3.md).
# ---------------------------------------------------------------------------


def pbkdf2(password: str, salt: bytes, dklen: int = 32) -> bytes:
    return hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        salt,
        PBKDF2_ITER,
        dklen=dklen,
    )


def hkdf_sha256(ikm: bytes, info: bytes, length: int,
                salt: bytes = b"") -> bytes:
    """RFC 5869 HKDF-SHA256, matching CryptoKit's empty-salt default."""
    if not salt:
        salt = b"\x00" * hashlib.sha256().digest_size
    prk = hmac.new(salt, ikm, hashlib.sha256).digest()
    out = b""
    t = b""
    counter = 1
    while len(out) < length:
        t = hmac.new(prk, t + info + bytes([counter]), hashlib.sha256).digest()
        out += t
        counter += 1
    return out[:length]


def keystream_hmac_ctr(key: bytes, info: bytes, length: int) -> bytes:
    """HMAC-SHA256 in counter mode -> arbitrary-length pseudo-random stream."""
    out = bytearray()
    counter = 0
    while len(out) < length:
        block = hmac.new(
            key,
            counter.to_bytes(4, "big") + info,
            hashlib.sha256,
        ).digest()
        out.extend(block)
        counter += 1
    return bytes(out[:length])


def derive_permutation(password: str, n: int) -> list[int]:
    """Fisher-Yates shuffle of [0, n) driven by HMAC-SHA256 in counter mode."""
    if n < 2:
        return list(range(n))
    perm_key = pbkdf2(password, PERM_KDF_SALT, dklen=32)
    ks = keystream_hmac_ctr(perm_key, PERM_INFO, n * 4)
    indices = list(range(n))
    i = n - 1
    while i > 0:
        off = i * 4
        r = int.from_bytes(ks[off:off + 4], "big")
        j = r % (i + 1)
        indices[i], indices[j] = indices[j], indices[i]
        i -= 1
    return indices


def decode_stealth(slots: Slots, password: str) -> bytes:
    available = slots.available_bytes
    plaintext_size = available - STEALTH_HEADER_SIZE - GCM_OVERHEAD
    if plaintext_size < STEALTH_LEN_PREFIX:
        raise ValueError("no payload found")

    permutation = derive_permutation(password, slots.n)
    stream = read_bytes(slots, available, order=permutation)

    salt = stream[0:16]
    marker = stream[16:20]
    masked_version = stream[20]
    masked_reserved = stream[21]

    main_key = pbkdf2(password, salt, dklen=32)
    derived = hkdf_sha256(main_key, HEADER_INFO, 34)
    mac_key = derived[0:32]
    mask = derived[32:34]

    expected = hmac.new(
        mac_key,
        salt + bytes([masked_version, masked_reserved]),
        hashlib.sha256,
    ).digest()[:4]
    if not hmac.compare_digest(expected, marker):
        raise ValueError("no payload found")

    version = masked_version ^ mask[0]
    if version != STEALTH_VERSION:
        raise ValueError("no payload found")

    cipher_blob = stream[STEALTH_HEADER_SIZE:available]
    aesgcm = AESGCM(main_key)
    try:
        plaintext_block = aesgcm.decrypt(
            cipher_blob[:GCM_NONCE_LEN],
            cipher_blob[GCM_NONCE_LEN:],
            None,
        )
    except Exception:
        raise ValueError("no payload found") from None

    if len(plaintext_block) < STEALTH_LEN_PREFIX:
        raise ValueError("payload truncated")
    length = int.from_bytes(plaintext_block[0:2], "big")
    if length == 0:
        raise ValueError("no payload found")
    if length > MAX_PAYLOAD_BYTES or length > len(plaintext_block) - STEALTH_LEN_PREFIX:
        raise ValueError("payload truncated")
    return plaintext_block[STEALTH_LEN_PREFIX:STEALTH_LEN_PREFIX + length]


# ---------------------------------------------------------------------------
# Top-level decode.
# ---------------------------------------------------------------------------


@dataclass
class DecodeResult:
    data: bytes
    encrypted: bool


def decode(png_path: Path, password: Optional[str] = None) -> DecodeResult:
    slots = load_slots(png_path)
    if password:
        try:
            return DecodeResult(data=decode_stealth(slots, password),
                                encrypted=True)
        except ValueError:
            pass    # fall through to open mode (e.g. legacy public sticker)
    return DecodeResult(data=decode_open(slots), encrypted=False)


# ---------------------------------------------------------------------------
# CLI.
# ---------------------------------------------------------------------------


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Reference decoder for Stiger PNG stickers.",
    )
    parser.add_argument("path", type=Path, help="Path to the sticker PNG.")
    parser.add_argument(
        "--password",
        type=str,
        default=None,
        help="Stealth-mode password. Omit for open-mode stickers.",
    )
    parser.add_argument(
        "--raw",
        action="store_true",
        help="Print payload as raw bytes to stdout instead of UTF-8 text.",
    )
    args = parser.parse_args(argv)

    try:
        result = decode(args.path, args.password)
    except ValueError as e:
        print(f"decode failed: {e}", file=sys.stderr)
        return 1

    mode = "stealth" if result.encrypted else "open"
    print(f"mode: {mode}", file=sys.stderr)
    if args.raw:
        sys.stdout.buffer.write(result.data)
    else:
        try:
            print(result.data.decode("utf-8"))
        except UnicodeDecodeError:
            sys.stdout.buffer.write(result.data)
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: reference/typescript/stiger_decode.ts
// Reference Decoder (TypeScript) — independent decoder for the Stiger wire
// format. Written from `../../stealth-v3.md`, `../../open-v1.md` and
// `../../lsb-layer.md` only. No Stiger code involved.
//
// Targets browsers and Node ≥ 22 (both expose globalThis.crypto.subtle).
// Mirrors the structure of the Python reference (`../python/stiger_decode.py`)
// so readers can compare the two side-by-side.

const subtle = globalThis.crypto.subtle;
const utf8 = new TextEncoder();

const PERM_SALT = utf8.encode('stiger-perm-v3-salt');
const PERM_INFO = utf8.encode('stiger-permutation-v3');
const HKDF_INFO = utf8.encode('stiger-v3-header');
const OPEN_MAGIC = new Uint8Array([0x53, 0x54, 0x45, 0x47]); // "STEG"

const OPEN_HEADER_SIZE = 9;
const OPEN_MAX_PAYLOAD = 2048;

const STEALTH_HEADER_SIZE = 22;
const STEALTH_GCM_OVERHEAD = 12 + 16;
const STEALTH_LEN_PREFIX = 2;
const STEALTH_VERSION = 0x03;
const STEALTH_MAX_MESSAGE = 2048;

export type DecodeMode = 'open-v1' | 'stealth-v3';

export type DecodeSuccess = {
  ok: true;
  mode: DecodeMode;
  message: string;
  bytes: Uint8Array;
};

export type DecodeFailure = {
  ok: false;
  reason:
    | 'no-payload'
    | 'too-small'
    | 'oversized'
    | 'invalid-utf8'
    | 'crypto-error';
  detail?: string;
};

export type DecodeResult = DecodeSuccess | DecodeFailure;

// ---------- WebCrypto wrappers ----------

async function pbkdf2(
  password: Uint8Array,
  salt: Uint8Array,
  iterations: number,
  lengthBytes: number,
): Promise<Uint8Array> {
  const baseKey = await subtle.importKey(
    'raw',
    password,
    { name: 'PBKDF2' },
    false,
    ['deriveBits'],
  );
  const bits = await subtle.deriveBits(
    { name: 'PBKDF2', hash: 'SHA-256', salt, iterations },
    baseKey,
    lengthBytes * 8,
  );
  return new Uint8Array(bits);
}

async function hmacSha256(
  key: Uint8Array,
  msg: Uint8Array,
): Promise<Uint8Array> {
  const k = await subtle.importKey(
    'raw',
    key,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  );
  const sig = await subtle.sign('HMAC', k, msg);
  return new Uint8Array(sig);
}

/**
 * HKDF-SHA256 with empty salt (RFC 5869). Matches CryptoKit's
 * `HKDF<SHA256>.deriveKey(inputKeyMaterial:info:outputByteCount:)` when
 * no salt is supplied — see stealth-v3 §4.
 *
 * RFC 5869 says: "if salt is not provided, it is set to a string of
 * HashLen zeros." We pass that explicit zero-filled buffer instead of
 * a zero-length Uint8Array — output is identical, and some WebCrypto
 * implementations (notably older iOS Safari) reject empty BufferSource.
 */
async function hkdfSha256Empty(
  ikm: Uint8Array,
  info: Uint8Array,
  lengthBytes: number,
): Promise<Uint8Array> {
  const baseKey = await subtle.importKey('raw', ikm, 'HKDF', false, ['deriveBits']);
  const bits = await subtle.deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt: new Uint8Array(32), info },
    baseKey,
    lengthBytes * 8,
  );
  return new Uint8Array(bits);
}

async function aesGcmOpen(
  key: Uint8Array,
  nonce: Uint8Array,
  ciphertextWithTag: Uint8Array,
): Promise<Uint8Array> {
  const k = await subtle.importKey('raw', key, { name: 'AES-GCM' }, false, ['decrypt']);
  const pt = await subtle.decrypt(
    { name: 'AES-GCM', iv: nonce, tagLength: 128 },
    k,
    ciphertextWithTag,
  );
  return new Uint8Array(pt);
}

function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i]! ^ b[i]!;
  return diff === 0;
}

function be32(n: number): Uint8Array {
  const out = new Uint8Array(4);
  out[0] = (n >>> 24) & 0xff;
  out[1] = (n >>> 16) & 0xff;
  out[2] = (n >>> 8) & 0xff;
  out[3] = n & 0xff;
  return out;
}

function readBe32(b: Uint8Array, o = 0): number {
  return ((b[o]! << 24) | (b[o + 1]! << 16) | (b[o + 2]! << 8) | b[o + 3]!) >>> 0;
}

function readBe16(b: Uint8Array, o = 0): number {
  return ((b[o]! << 8) | b[o + 1]!) >>> 0;
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.length;
  }
  return out;
}

// ---------- LSB transport (lsb-layer.md) ----------

/** Eligible-pixel indices (alpha === 0xFF), in row-major order. */
function eligibleIndices(rgba: Uint8ClampedArray | Uint8Array): Uint32Array {
  const total = rgba.length / 4;
  const out: number[] = [];
  for (let i = 0; i < total; i++) {
    if (rgba[i * 4 + 3] === 0xff) out.push(i);
  }
  return new Uint32Array(out);
}

/** Read `byteCount` MSB-first bytes through the slot map. */
function readBytes(
  rgba: Uint8ClampedArray | Uint8Array,
  eligibles: Uint32Array,
  slotMap: Uint32Array,
  startSlot: number,
  byteCount: number,
): Uint8Array {
  const out = new Uint8Array(byteCount);
  for (let i = 0; i < byteCount; i++) {
    let b = 0;
    for (let k = 0; k < 8; k++) {
      const logical = startSlot + i * 8 + k;
      const pixelIdx = eligibles[slotMap[logical]!]!;
      const bit = rgba[pixelIdx * 4 + 2]! & 1;
      b |= bit << (7 - k);
    }
    out[i] = b;
  }
  return out;
}

function rowMajorSlotMap(n: number): Uint32Array {
  const out = new Uint32Array(n);
  for (let i = 0; i < n; i++) out[i] = i;
  return out;
}

// ---------- Permutation π (stealth-v3 §3) ----------

async function keystream(permKey: Uint8Array, lengthBytes: number): Promise<Uint8Array> {
  const out = new Uint8Array(lengthBytes);
  let off = 0;
  let counter = 0;
  while (off < lengthBytes) {
    const block = await hmacSha256(permKey, concat(be32(counter), PERM_INFO));
    const take = Math.min(block.length, lengthBytes - off);
    out.set(block.subarray(0, take), off);
    off += take;
    counter += 1;
  }
  return out;
}

async function permutation(password: Uint8Array, N: number): Promise<Uint32Array> {
  const permKey = await pbkdf2(password, PERM_SALT, 600_000, 32);
  const ks = await keystream(permKey, N * 4);
  const indices = new Uint32Array(N);
  for (let i = 0; i < N; i++) indices[i] = i;
  // Fisher–Yates: i = N-1 down to 1, j = uint32_be(ks[i*4..]) mod (i+1)
  for (let i = N - 1; i >= 1; i--) {
    const r = readBe32(ks, i * 4);
    const j = r % (i + 1);
    if (j !== i) {
      const tmp = indices[i]!;
      indices[i] = indices[j]!;
      indices[j] = tmp;
    }
  }
  return indices;
}

// ---------- Public API ----------

/**
 * Decode a Stiger PNG given its straight (non-premultiplied) RGBA pixels.
 * Tries open-v1 first; tries stealth-v3 only if a non-empty password is
 * provided. Wrong password is indistinguishable from "no payload" by design.
 */
export async function decode(
  rgba: Uint8ClampedArray | Uint8Array,
  password?: string,
): Promise<DecodeResult> {
  const eligibles = eligibleIndices(rgba);
  const N = eligibles.length;
  const availableBytes = Math.floor(N / 8);

  const openResult = tryOpen(rgba, eligibles, availableBytes);
  if (openResult.ok) return openResult;

  if (password && password.length > 0) {
    return tryStealth(rgba, eligibles, password);
  }

  return {
    ok: false,
    reason: 'no-payload',
    detail: 'open-v1 magic not present, no password provided for stealth-v3',
  };
}

// ---------- open-v1 (open-v1.md) ----------

function tryOpen(
  rgba: Uint8ClampedArray | Uint8Array,
  eligibles: Uint32Array,
  availableBytes: number,
): DecodeResult {
  if (availableBytes < OPEN_HEADER_SIZE) return { ok: false, reason: 'too-small' };
  const slotMap = rowMajorSlotMap(eligibles.length);
  const header = readBytes(rgba, eligibles, slotMap, 0, OPEN_HEADER_SIZE);

  for (let i = 0; i < 4; i++) {
    if (header[i] !== OPEN_MAGIC[i]) {
      return { ok: false, reason: 'no-payload', detail: 'open magic mismatch' };
    }
  }
  if (header[4] !== 0x01) {
    return { ok: false, reason: 'no-payload', detail: `open version ${header[4]} unsupported` };
  }
  const payloadLength = readBe32(header, 5);
  if (payloadLength === 0) {
    return { ok: false, reason: 'no-payload', detail: 'open payloadLength == 0' };
  }
  if (payloadLength > OPEN_MAX_PAYLOAD) {
    return { ok: false, reason: 'oversized', detail: 'open payloadLength > 2048' };
  }
  const cap = availableBytes - OPEN_HEADER_SIZE;
  if (payloadLength > cap) {
    return { ok: false, reason: 'oversized', detail: 'open payloadLength > capacity' };
  }

  const payload = readBytes(rgba, eligibles, slotMap, OPEN_HEADER_SIZE * 8, payloadLength);
  return { ok: true, mode: 'open-v1', message: utf8Lossy(payload), bytes: payload };
}

// ---------- stealth-v3 (stealth-v3.md) ----------

async function tryStealth(
  rgba: Uint8ClampedArray | Uint8Array,
  eligibles: Uint32Array,
  password: string,
): Promise<DecodeResult> {
  const N = eligibles.length;
  const availableBytes = Math.floor(N / 8);
  const plaintextSize = availableBytes - STEALTH_HEADER_SIZE - STEALTH_GCM_OVERHEAD;
  if (plaintextSize < STEALTH_LEN_PREFIX) {
    return { ok: false, reason: 'too-small', detail: 'carrier too small for stealth-v3' };
  }

  const pwdBytes = utf8.encode(password);

  // Permutation π over [0, N), keyed by password alone (no per-message salt).
  const pi = await permutation(pwdBytes, N);

  // Read the 22-byte header through π.
  const header = readBytes(rgba, eligibles, pi, 0, STEALTH_HEADER_SIZE);
  const salt = header.subarray(0, 16);
  const marker = header.subarray(16, 20);
  const maskedVersion = header[20]!;
  const maskedReserved = header[21]!;

  // Per-message keys.
  const mainKey = await pbkdf2(pwdBytes, salt, 600_000, 32);
  const derived = await hkdfSha256Empty(mainKey, HKDF_INFO, 34);
  const macKey = derived.subarray(0, 32);
  const mask = derived.subarray(32, 34);

  // Constant-time marker verification.
  const macInput = concat(salt, new Uint8Array([maskedVersion, maskedReserved]));
  const expected = (await hmacSha256(macKey, macInput)).subarray(0, 4);
  if (!constantTimeEqual(expected, marker)) {
    return {
      ok: false,
      reason: 'no-payload',
      detail: 'marker mismatch (wrong password or not a stealth carrier)',
    };
  }

  const version = (maskedVersion ^ mask[0]!) & 0xff;
  if (version !== STEALTH_VERSION) {
    return { ok: false, reason: 'no-payload', detail: `stealth version ${version} unsupported` };
  }

  // GCM open the rest of the LSB stream.
  const gcmBlobLen = availableBytes - STEALTH_HEADER_SIZE;
  const gcmBlob = readBytes(rgba, eligibles, pi, STEALTH_HEADER_SIZE * 8, gcmBlobLen);
  const nonce = gcmBlob.subarray(0, 12);
  const ctWithTag = gcmBlob.subarray(12);

  let plaintext: Uint8Array;
  try {
    plaintext = await aesGcmOpen(mainKey, nonce, ctWithTag);
  } catch {
    return { ok: false, reason: 'no-payload', detail: 'GCM authentication failed' };
  }

  if (plaintext.length < STEALTH_LEN_PREFIX) {
    return { ok: false, reason: 'no-payload', detail: 'plaintext shorter than length prefix' };
  }
  const length = readBe16(plaintext, 0);
  if (length === 0) return { ok: false, reason: 'no-payload', detail: 'plaintext length == 0' };
  if (length > plaintext.length - STEALTH_LEN_PREFIX) {
    return { ok: false, reason: 'oversized', detail: 'plaintext length exceeds plaintextSize - 2' };
  }
  if (length > STEALTH_MAX_MESSAGE) {
    return { ok: false, reason: 'oversized', detail: 'plaintext length > 2048' };
  }

  const messageBytes = plaintext.subarray(STEALTH_LEN_PREFIX, STEALTH_LEN_PREFIX + length);
  return {
    ok: true,
    mode: 'stealth-v3',
    message: utf8Lossy(messageBytes),
    bytes: messageBytes,
  };
}

function utf8Lossy(bytes: Uint8Array): string {
  return new TextDecoder('utf-8', { fatal: false }).decode(bytes);
}
